### PR TITLE
Use go version 1.21.0 in go.mod directives

### PR DESCRIPTION
**Pull request by liustanley**
### What does this PR do?

This PR changes the go version from 1.21.9 to 1.21.0 in all go.mod directives.

### Motivation

The go version upgrade to 1.21.9 was a breaking change for our components in opentelemetry-collector-contrib which still uses go 1.21 in their go.mod files. When we try to upgrade the datadog-agent dependencies in the collector, this also changes the collector go.mod version to 1.21.9 ([example](https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/32327/files#r1561492954)).

opentelemetry-collector-contrib uses a majority of the datadog-agent modules, so we decided to use `go 1.21.0` for all of the datadog-agent modules in order to avoid future conflicts.

See slack thread for more context: https://dd.slack.com/archives/C03ER8VH3NY/p1713189823890899

### CODE REVIEW

File: pkg/security/secl/go.mod [hunk 1]
```diff
@@ -1,6 +1,6 @@
 module github.com/DataDog/datadog-agent/pkg/security/secl
 
-go 1.21
+go 1.21.0
```
Comment (paulcacheux) [inline pkg/security/secl/go.mod:3]: Please don't, `go 1.21` is done on purpose here

File: pkg/security/seclwin/go.mod [hunk 2]
```diff
@@ -1,6 +1,6 @@
 module github.com/DataDog/datadog-agent/pkg/security/seclwin
 
-go 1.21
+go 1.21.0
```
Comment (paulcacheux) [inline pkg/security/seclwin/go.mod:3]: Same here

**Review by paulcacheux — CHANGES_REQUESTED** [review]

**Review by paulcacheux — APPROVED** [review]

**Review by hush-hush — APPROVED** [review]

**Comment by raphaelgavache** [comment]
For fleet part there's an issue due to a file rename, I've prepared you a quick fix for you: https://github.com/DataDog/datadog-agent/pull/24721

File: tasks/update_go.py [hunk 3]
```diff
@@ -191,14 +191,9 @@ def _update_go_mods(warn: bool, version: str, include_otel_modules: bool, dry_ru
...
-        if module.legacy_go_mod_version:
-            major_minor = _get_major_minor_version(version)
-            # $ only matches \n, not \r\n, so we need to use \r?$ to make it work on Windows
-            _update_file(warn, mod_file, f"^go {PATTERN_MAJOR_MINOR}\r?$", f"go {major_minor}", dry_run=dry_run)
-        else:
-            # $ only matches \n, not \r\n, so we need to use \r?$ to make it work on Windows
-            _update_file(warn, mod_file, f"^go {PATTERN_MAJOR_MINOR_BUGFIX}\r?$", f"go {version}", dry_run=dry_run)
+        major_minor = _get_major_minor_version(version)
+        # $ only matches \n, not \r\n, so we need to use \r?$ to make it work on Windows
+        _update_file(warn, mod_file, f"^go {PATTERN_MAJOR_MINOR}\r?$", f"go {major_minor}", dry_run=dry_run)
```
[the remark this answers is not in the document]
Comment (pgimalac) [reply tasks/update_go.py:196]: Thanks for updating this btw !

File: tasks/update_go.py [hunk 4]
```diff
@@ -191,14 +191,16 @@ def _update_go_mods(warn: bool, version: str, include_otel_modules: bool, dry_ru
...
             continue
         mod_file = f"./{path}/go.mod"
-
+        major_minor = _get_major_minor_version(version)
         if module.legacy_go_mod_version:
-            major_minor = _get_major_minor_version(version)
             # $ only matches \n, not \r\n, so we need to use \r?$ to make it work on Windows
             _update_file(warn, mod_file, f"^go {PATTERN_MAJOR_MINOR}\r?$", f"go {major_minor}", dry_run=dry_run)
         else:
+            major_minor_zero = f"{major_minor}.0"
```
Comment (FlorentClarret) [inline tasks/update_go.py:199]: `major_minor` might not be initialized

Comment (FlorentClarret) [reply tasks/update_go.py:199]: oh nvm, you just fixed it

**Review by CelianR — APPROVED** [review]

**Review by FlorentClarret — APPROVED** [review]

**Review by mx-psi — APPROVED** [review]

**Review by knusbaum — APPROVED** [review]

**Review by rayz — APPROVED** [review]

**Review by robertjli — APPROVED** [review]

**Review by pducolin — APPROVED** [review]
Ok for Agent Developer Tools

**Review by pgimalac — APPROVED** [review]

**Review by jszwedko — APPROVED** [review]

**Review by zoedt — APPROVED** [review]

**Review by gh123man — APPROVED** [review]

**Review by clarkb7 — APPROVED** [review]

**Review by raphaelgavache — APPROVED** [review]

**Review by BaptisteFoy — APPROVED** [review]

**Review by davidor — APPROVED** [review]

**Review by mfpierre — APPROVED** [review]

**Pull request merged**